Realm/ObjectServerTests/RLMCollectionSyncTests.mm

```diff
@@ -0,0 +1,411 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMSyncTestCase.h"
+
+#if TARGET_OS_OSX
+
+#pragma mark RLMSet Sync Tests
+
+@interface RLMSetObjectServerTests : RLMSyncTestCase
+@end
+
+@implementation RLMSetObjectServerTests
+
+- (void)roundTripWithKeyPath:(NSString *)keyPath
+                      values:(NSArray *)values
+                otherKeyPath:(NSString *)otherKeyPath
+                 otherValues:(NSArray *)otherValues
+                    isObject:(BOOL)isObject
+                  callerName:(NSString *)callerName {
+    try {
+        RLMUser *user = [self logInUserForCredentials:[self basicCredentialsWithName:callerName
+                                                                            register:self.isParent]];
+        RLMRealm *realm = [self openRealmForPartitionValue:NSStringFromSelector(_cmd) user:user];
+
+        if (self.isParent) {
+            [realm transactionWithBlock:^{
+                [realm deleteAllObjects];
+            }];
+            [self waitForDownloadsForRealm:realm];
+            CHECK_COUNT(0, RLMSetSyncObject, realm);
+            RLMRunChildAndWait();
+            [self waitForDownloadsForRealm:realm];
+            CHECK_COUNT(1, RLMSetSyncObject, realm);
+            // Run the child again to add the values
+            RLMRunChildAndWait();
+            [self waitForDownloadsForRealm:realm];
+            CHECK_COUNT(1, RLMSetSyncObject, realm);
+            RLMResults<RLMSetSyncObject *> *results
+            = [RLMSetSyncObject allObjectsInRealm:realm];
+            RLMSetSyncObject *obj = results.firstObject;
+
+            XCTAssertEqual(((RLMSet *)obj[keyPath]).count, values.count);
+            XCTAssertEqual(((RLMSet *)obj[otherKeyPath]).count, otherValues.count);
+            // Run the child again to intersect the values
+            RLMRunChildAndWait();
+            [self waitForDownloadsForRealm:realm];
+            CHECK_COUNT(1, RLMSetSyncObject, realm);
+
+            if (!isObject) {
+                XCTAssertTrue([((RLMSet *)obj[keyPath]) intersectsSet:((RLMSet *)obj[otherKeyPath])]);
```

Comment:
Reading properties requires casting to the type, but calling methods doesn't (and it's not even really useful as documentation because casting to an unrelated type which happens to have the same method works).

Reply:
addressed this in #7190